src/main/java/se/sundsvall/billingpreprocessor/service/InvoiceFileService.java

```diff
@@ -0,0 +1,128 @@
+package se.sundsvall.billingpreprocessor.service;
+
+import static se.sundsvall.billingpreprocessor.integration.db.model.enums.Status.APPROVED;
+import static se.sundsvall.billingpreprocessor.integration.db.model.enums.Status.INVOICED;
+import static se.sundsvall.billingpreprocessor.service.mapper.InvoiceFileMapper.toInvoiceFileEntity;
+
+import java.io.ByteArrayOutputStream;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import jakarta.transaction.Transactional;
+import se.sundsvall.billingpreprocessor.integration.db.BillingRecordRepository;
+import se.sundsvall.billingpreprocessor.integration.db.InvoiceFileRepository;
+import se.sundsvall.billingpreprocessor.integration.db.model.BillingRecordEntity;
+import se.sundsvall.billingpreprocessor.integration.db.model.enums.Type;
+import se.sundsvall.billingpreprocessor.service.creator.CreationError;
+import se.sundsvall.billingpreprocessor.service.creator.InvoiceCreator;
+
+@Service
+public class InvoiceFileService {
```

Comment:
- **createFileEntities** is a method for creating file entities. It will be called by a scheduled job and there is no point in returning anything. Methods can be named create... even though they are not returning things.
- **processBillingRecords** -> returns a list of errors to be handled, if any errors occurs during record processing
- **createBillingFile** -> each billing record will become a part of a billing file, the list returned from the method is part of the total list of errors that is returned from calling method **processBillingRecords**
- **createBillingRecord** -> As each billing file contains one or more billing records, and each billing record creation can result in an error, there is a need for returning an optional error object (which will be added to the list of errors if present)

One of the purposes of the interface is to simplify the code by not needing to wire in each creator, but instead just wiring in a list of interfaces. Also, the service class does not need to have logic for knowing which creator to call depending on data in the billing record entry, it just need to use the methods defined in the interface. If we need to add a new creator (i,e. create a new kind of billing file) in the future, all we need to do is to define a new creator and let it implement the methods of the interface. No modifications to the service will be needed.

Belive me, I tried to make it as simple as it could be, but sometimes the reallity is more complex than one would wish. If you have any suggestions for improvements and still contain the needed functionality, I'm all ears.